benchmark/main.go

```diff
@@ -0,0 +1,342 @@
+		results[i] = <-resCh
+	}
+
+	fmt.Println("processing results")
+	totalTime := time.Now().Sub(start)
+	totals := calculateTotalResults(results, totalTime, &subTimes)
+	if totals == nil {
+		return
+	}
+	// print stats
+	printResults(results, totals, *format)
+}
+func calculateTotalResults(results []*RunResults, totalTime time.Duration, subTimes *SubTimes) *TotalResults {
+	if results == nil || len(results) < 1 {
+		return nil
+	}
+	totals := new(TotalResults)
+	totals.TotalRunTime = totalTime.Seconds()
+	var subTimeRunResults RunResults
+	msgTimeMeans := make([]float64, len(results))
+	msgTimeMeansDelivered := make([]float64, len(results))
+	msgsPerSecs := make([]float64, len(results))
+	runTimes := make([]float64, len(results))
+	bws := make([]float64, len(results))
+
+	totals.MsgTimeMin = results[0].MsgTimeMin
+	for i, res := range results {
+
+		times := (*subTimes)[res.ID]
+		subTimeRunResults.MsgTimeMin = stats.StatsMin(times)
+		subTimeRunResults.MsgTimeMax = stats.StatsMax(times)
+		subTimeRunResults.MsgTimeMean = stats.StatsMean(times)
+		subTimeRunResults.MsgTimeStd = stats.StatsSampleStandardDeviation(times)
+
+		res.MsgDelTimeMin = subTimeRunResults.MsgTimeMin
+		res.MsgDelTimeMax = subTimeRunResults.MsgTimeMax
+		res.MsgDelTimeMean = subTimeRunResults.MsgTimeMean
+		res.MsgDelTimeStd = subTimeRunResults.MsgTimeStd
+
+		totals.Successes += res.Successes
+		totals.Failures += res.Failures
+		totals.TotalMsgsPerSec += res.MsgsPerSec
+
+		if res.MsgTimeMin < totals.MsgTimeMin {
+			totals.MsgTimeMin = res.MsgTimeMin
+		}
+
+		if res.MsgTimeMax > totals.MsgTimeMax {
+			totals.MsgTimeMax = res.MsgTimeMax
+		}
+
+		if subTimeRunResults.MsgTimeMin < totals.MsgDelTimeMin {
+			totals.MsgDelTimeMin = subTimeRunResults.MsgTimeMin
+		}
+
+		if subTimeRunResults.MsgTimeMax > totals.MsgDelTimeMax {
+			totals.MsgDelTimeMax = subTimeRunResults.MsgTimeMax
+		}
+
+		msgTimeMeansDelivered[i] = subTimeRunResults.MsgTimeMean
+		msgTimeMeans[i] = res.MsgTimeMean
+		msgsPerSecs[i] = res.MsgsPerSec
+		runTimes[i] = res.RunTime
+		bws[i] = res.MsgsPerSec
+	}
+	totals.Ratio = float64(totals.Successes) / float64(totals.Successes+totals.Failures)
+	totals.AvgMsgsPerSec = stats.StatsMean(msgsPerSecs)
+	totals.AvgRunTime = stats.StatsMean(runTimes)
+	totals.MsgDelTimeMeanAvg = stats.StatsMean(msgTimeMeansDelivered)
+	totals.MsgDelTimeMeanStd = stats.StatsSampleStandardDeviation(msgTimeMeansDelivered)
+	totals.MsgTimeMeanAvg = stats.StatsMean(msgTimeMeans)
+	totals.MsgTimeMeanStd = stats.StatsSampleStandardDeviation(msgTimeMeans)
+
+	return totals
+}
+
+func printResults(results []*RunResults, totals *TotalResults, format string) {
+	switch format {
+	case "json":
+		jr := JSONResults{
+			Runs:   results,
+			Totals: totals,
+		}
+		data, _ := json.Marshal(jr)
+		var out bytes.Buffer
+		json.Indent(&out, data, "", "\t")
+
+		fmt.Println(string(out.Bytes()))
+	default:
+		for _, res := range results {
+			fmt.Printf("======= CLIENT %s =======\n", res.ID)
```

Comment:
You can use [Aurora](https://github.com/logrusorgru/aurora) package to add some colors in the CLI and make is sexier ;).